dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/icon/IconController.java

```diff
@@ -94,102 +96,135 @@ public class IconController {
 
   private final FileResourceService fileResourceService;
 
-  private final IconMapper iconMapper;
-
   private final DhisConfigurationProvider dhisConfig;
 
+  private final RenderService renderService;
+
+  private final IconMapper iconMapper;
+
   private final FieldFilterService fieldFilterService;
 
-  private final IconRequestParamsMapper iconRequestParamsMapper;
+  private final ContextService contextService;
 
   private final LinkService linkService;
 
-  @GetMapping("/{iconKey}")
-  public @ResponseBody IconResponse getIcon(@PathVariable String iconKey) throws NotFoundException {
-    Icon icon = iconService.getIcon(iconKey);
+  private final IconRequestParamMapper iconRequestParamMapper;
 
-    return iconMapper.from(icon);
-  }
+  @GetMapping
+  public @ResponseBody PaginatedIconResponse getAllIcons(IconRequestParams iconRequestParams)
+      throws BadRequestException {
+    IconQueryParams iconQueryParams = iconRequestParamMapper.map(iconRequestParams);
 
-  @GetMapping("/{iconKey}/icon.svg")
-  @Deprecated
-  public void getIconData(
-      HttpServletResponse response, HttpServletRequest request, @PathVariable String iconKey)
-      throws IOException {
-    String location = response.encodeRedirectURL("/icons/" + iconKey + "/icon");
-    response.sendRedirect(ContextUtils.getRootPath(request) + location);
+    Pager pager = null;
+
+    if (iconRequestParams.isPaging()) {
+      long total = iconService.count(iconQueryParams);
+      pager = new Pager(iconRequestParams.getPage(), total, iconRequestParams.getPageSize());
+      iconQueryParams.setPager(pager);
+    }
+
+    List<Icon> icons = iconService.getIcons(iconQueryParams);
+
+    icons.forEach(
+        i ->
+            i.setHref(
+                i.isCustom()
+                    ? getCustomIconReference(i.getKey())
+                    : getDefaultIconReference(i.getKey())));
+
+    List<ObjectNode> objectNodes =
+        fieldFilterService.toObjectNodes(icons.stream().toList(), iconRequestParams.getFields());
+
+    linkService.generatePagerLinks(pager, Icon.class);
+
+    return new PaginatedIconResponse(pager, objectNodes);
   }
 
   @GetMapping(value = "/{key}/icon")
   public void getIconData(@PathVariable String key, HttpServletResponse response)
       throws NotFoundException, WebMessageException, IOException {
-
     Icon icon = iconService.getIcon(key);
```

Comment:
This route is hit every time an icon is requested as a static asset. Is this something we should consider performance wise? Could it be considered bad-performance wise to instantiate and grab all the data from the DB, when we really only need to check for the existence of `key` and get the `fileResource`. Or is this not a problem?

@teleivo

Reply:
Icons have a max age of 1 year. The icons are served from the browsers cache if I open capture on dev. So unless they are not updated (or removed from the cache) icons should not hit the backend more than once / year / user.